File: src/lib/confidence.ts
import { PingLog } from "@/types";
import { DATA_FRESHNESS } from "./constants";
import { calculateJitter } from "./latency";

/**
 * Confidence Algorithm
 *
 * Score = (Unique pings in area) × (1 / StdDev of Latency)
 *
 * Data Freshness Weights:
 *   < 7 days:  1.0
 *   7-30 days: 0.5
 *   > 30 days: ignored
 */

export type ConfidenceLevel = "high" | "medium" | "low";

export interface ConfidenceResult {
  score: number;
  topISP: string;
  confidenceLevel: ConfidenceLevel;
}

/**
 * Get the freshness weight for a ping based on its age
 */
export function getFreshnessWeight(timestamp: number): number {
  const now = Date.now();
  const ageMs = now - timestamp;
  const ageDays = ageMs / (1000 * 60 * 60 * 24);

  if (ageDays <= DATA_FRESHNESS.FRESH_DAYS) {
    return 1.0;
  } else if (ageDays <= DATA_FRESHNESS.STALE_DAYS) {
    return 0.5;
  } else {
    return 0; // Expired - ignore
  }
}

/**
 * Filter pings to only include fresh ones (not expired)
 */
export function filterFreshPings(pings: PingLog[]): PingLog[] {
  return pings.filter((ping) => getFreshnessWeight(ping.timestamp) > 0);
}

/**
 * Calculate weighted average latency based on data freshness
 */
export function calculateWeightedAverageLatency(pings: PingLog[]): number {
  const freshPings = filterFreshPings(pings);
  if (freshPings.length === 0) return 0;

  let totalWeight = 0;
  let weightedSum = 0;

  freshPings.forEach((ping) => {
    const weight = getFreshnessWeight(ping.timestamp);
    weightedSum += ping.latencyMs * weight;
    totalWeight += weight;
  });

  return totalWeight > 0 ? Math.round(weightedSum / totalWeight) : 0;
}

/**
 * Get confidence level based on score
 */
export function getConfidenceLevel(score: number): ConfidenceLevel {
  if (score >= 70) return "high";
  if (score >= 40) return "medium";
  return "low";
}

/**
 * Calculate the confidence score for a set of pings
 *
 * Higher score = more reliable data
 * Score considers: number of pings, consistency (low std dev), freshness
 */
export function calculateConfidenceScore(pings: PingLog[]): number {
  const freshPings = filterFreshPings(pings);
  if (freshPings.length === 0) return 0;

  // Factor 1: Number of unique pings (weighted by freshness)
  let weightedCount = 0;
  freshPings.forEach((ping) => {
    weightedCount += getFreshnessWeight(ping.timestamp);
  });

  // Factor 2: Consistency (inverse of standard deviation)
  const latencies = freshPings.map((p) => p.latencyMs);
  const jitter = calculateJitter(latencies);
  // Avoid division by zero; add 1 to jitter
  const consistencyFactor = 1 / (jitter + 1);

  // Combine factors
  // Normalize: more pings and lower jitter = higher score
  const rawScore = weightedCount * consistencyFactor * 10;

  // Cap at 100 for display purposes
  return Math.min(100, Math.round(rawScore));
}

/**
 * Calculate full confidence result with score, topISP, and level
 */
export function calculateConfidence(pings: PingLog[]): ConfidenceResult {
  const score = calculateConfidenceScore(pings);
  const topISP = getTopISP(pings);
  const confidenceLevel = getConfidenceLevel(score);

  return {
    score,
    topISP,
    confidenceLevel,
  };
}

/**
 * Calculate the consistency percentage
 * Returns the percentage of pings that have "good" latency
 */
export function calculateConsistency(
  pings: PingLog[],
  threshold: number = 100
): number {
  const freshPings = filterFreshPings(pings);
  if (freshPings.length === 0) return 0;

  const goodPings = freshPings.filter((p) => p.latencyMs <= threshold);
  return Math.round((goodPings.length / freshPings.length) * 100);
}

/**
 * Get the most common ISP in a set of pings
 * Tie-breaker: If two ISPs have the same count, pick the one with lower avg latency
 */
export function getTopISP(pings: PingLog[]): string {
  const freshPings = filterFreshPings(pings);
  if (freshPings.length === 0) return "No Data";

  // Track count and total latency for each ISP
  const ispStats = new Map<string, { count: number; totalLatency: number }>();

  freshPings.forEach((ping) => {
    const existing = ispStats.get(ping.reportedISP);
    if (existing) {
      existing.count += 1;
      existing.totalLatency += ping.latencyMs;
    } else {
      ispStats.set(ping.reportedISP, {
        count: 1,
        totalLatency: ping.latencyMs,
      });
    }
  });

  let topISP = "No Data";
  let maxCount = 0;
  let bestAvgLatency = Infinity;

  ispStats.forEach((stats, isp) => {
    const avgLatency = stats.totalLatency / stats.count;

    // Higher count wins, or same count with lower latency wins
    if (
      stats.count > maxCount ||
      (stats.count === maxCount && avgLatency < bestAvgLatency)
    ) {
      maxCount = stats.count;
      bestAvgLatency = avgLatency;
      topISP = isp;
    }
  });

  return maxCount > 0 ? topISP : "No Data";
}

/**
 * Get ISP rankings sorted by average latency (fastest first)
 */
export function getISPRankings(
  pings: PingLog[]
): Array<{ isp: string; avgLatency: number; count: number }> {
  const freshPings = filterFreshPings(pings);
  if (freshPings.length === 0) return [];

  const ispData = new Map<string, { total: number; count: number }>();

  freshPings.forEach((ping) => {
    const existing = ispData.get(ping.reportedISP);
    if (existing) {
      existing.total += ping.latencyMs;
      existing.count += 1;
    } else {
      ispData.set(ping.reportedISP, {
        total: ping.latencyMs,
        count: 1,
      });
    }
  });

  const rankings: Array<{ isp: string; avgLatency: number; count: number }> = [];

  ispData.forEach((data, isp) => {
    rankings.push({
      isp,
      avgLatency: Math.round(data.total / data.count),
      count: data.count,
    });
  });

  // Sort by average latency (ascending - fastest first)
  return rankings.sort((a, b) => a.avgLatency - b.avgLatency);
}

File: apps/mobile/lib/domain/confidence.dart
import 'dart:math' as math;

import 'constants.dart';
import 'entities/ping_log.dart';
import 'latency.dart';

/// Confidence Algorithm
///
/// Score = (Unique pings in area) x (1 / StdDev of Latency)
///
/// Data Freshness Weights:
///   < 7 days:  1.0
///   7-30 days: 0.5
///   > 30 days: ignored
enum ConfidenceLevel { high, medium, low }

class ConfidenceResult {
  const ConfidenceResult({
    required this.score,
    required this.topISP,
    required this.confidenceLevel,
  });

  final int score;
  final String topISP;
  final ConfidenceLevel confidenceLevel;
}

class IspRanking {
  const IspRanking({required this.isp, required this.avgLatency, required this.count});

  final String isp;
  final int avgLatency;
  final int count;
}

double getFreshnessWeight(DateTime timestamp) {
  final ageMs = DateTime.now().difference(timestamp).inMilliseconds;
  final ageDays = ageMs / (1000 * 60 * 60 * 24);

  if (ageDays <= DataFreshness.freshDays) {
    return 1.0;
  } else if (ageDays <= DataFreshness.staleDays) {
    return 0.5;
  } else {
    return 0;
  }
}

List<PingLog> filterFreshPings(List<PingLog> pings) {
  return pings.where((ping) => getFreshnessWeight(ping.timestamp) > 0).toList();
}

int calculateWeightedAverageLatency(List<PingLog> pings) {
  final freshPings = filterFreshPings(pings);
  if (freshPings.isEmpty) return 0;

  var totalWeight = 0.0;
  var weightedSum = 0.0;

  for (final ping in freshPings) {
    final weight = getFreshnessWeight(ping.timestamp);
    weightedSum += ping.latencyMs * weight;
    totalWeight += weight;
  }

  return totalWeight > 0 ? (weightedSum / totalWeight).round() : 0;
}

ConfidenceLevel getConfidenceLevel(int score) {
  if (score >= 70) return ConfidenceLevel.high;
  if (score >= 40) return ConfidenceLevel.medium;
  return ConfidenceLevel.low;
}

/// Calculate the confidence score for a set of pings.
///
/// Higher score = more reliable data. Score considers: number of pings,
/// consistency (low std dev), freshness.
int calculateConfidenceScore(List<PingLog> pings) {
  final freshPings = filterFreshPings(pings);
  if (freshPings.isEmpty) return 0;

  var weightedCount = 0.0;
  for (final ping in freshPings) {
    weightedCount += getFreshnessWeight(ping.timestamp);
  }

  final latencies = freshPings.map((p) => p.latencyMs).toList();
  final jitter = calculateJitter(latencies);
  final consistencyFactor = 1 / (jitter + 1);

  final rawScore = weightedCount * consistencyFactor * 10;

  return math.min(100, rawScore.round());
}

ConfidenceResult calculateConfidence(List<PingLog> pings) {
  final score = calculateConfidenceScore(pings);
  final topISP = getTopISP(pings);
  final confidenceLevel = getConfidenceLevel(score);

  return ConfidenceResult(score: score, topISP: topISP, confidenceLevel: confidenceLevel);
}

/// Percentage of fresh pings with latency at or under [threshold].
int calculateConsistency(List<PingLog> pings, [int threshold = 100]) {
  final freshPings = filterFreshPings(pings);
  if (freshPings.isEmpty) return 0;

  final goodPings = freshPings.where((p) => p.latencyMs <= threshold);
  return (goodPings.length / freshPings.length * 100).round();
}

class _IspStats {
  _IspStats(this.count, this.totalLatency);
  int count;
  int totalLatency;
}

/// Most common ISP among fresh pings. Tie-breaker: equal counts resolve to
/// the lower average latency.
String getTopISP(List<PingLog> pings) {
  final freshPings = filterFreshPings(pings);
  if (freshPings.isEmpty) return 'No Data';

  final ispStats = <String, _IspStats>{};

  for (final ping in freshPings) {
    final existing = ispStats[ping.reportedIsp];
    if (existing != null) {
      existing.count += 1;
      existing.totalLatency += ping.latencyMs;
    } else {
      ispStats[ping.reportedIsp] = _IspStats(1, ping.latencyMs);
    }
  }

  var topISP = 'No Data';
  var maxCount = 0;
  var bestAvgLatency = double.infinity;

  ispStats.forEach((isp, stats) {
    final avgLatency = stats.totalLatency / stats.count;

    if (stats.count > maxCount || (stats.count == maxCount && avgLatency < bestAvgLatency)) {
      maxCount = stats.count;
      bestAvgLatency = avgLatency;
      topISP = isp;
    }
  });

  return maxCount > 0 ? topISP : 'No Data';
}

class _IspTotals {
  _IspTotals(this.total, this.count);
  int total;
  int count;
}

/// ISP rankings sorted by average latency (fastest first).
List<IspRanking> getISPRankings(List<PingLog> pings) {
  final freshPings = filterFreshPings(pings);
  if (freshPings.isEmpty) return [];

  final ispData = <String, _IspTotals>{};

  for (final ping in freshPings) {
    final existing = ispData[ping.reportedIsp];
    if (existing != null) {
      existing.total += ping.latencyMs;
      existing.count += 1;
    } else {
      ispData[ping.reportedIsp] = _IspTotals(ping.latencyMs, 1);
    }
  }

  final rankings = <IspRanking>[
    for (final entry in ispData.entries)
      IspRanking(
        isp: entry.key,
        avgLatency: (entry.value.total / entry.value.count).round(),
        count: entry.value.count,
      ),
  ];

  rankings.sort((a, b) => a.avgLatency - b.avgLatency);
  return rankings;
}
